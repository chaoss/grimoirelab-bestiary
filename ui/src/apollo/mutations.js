import gql from "graphql-tag";

const ADD_PROJECT = gql`
  mutation addProject(
    $ecosystemId: ID
    $name: String
    $title: String
    $parentId: ID
  ) {
    addProject(
      ecosystemId: $ecosystemId
      name: $name
      title: $title
      parentId: $parentId
    ) {
      project {
        id
        name
      }
    }
  }
`;

const DELETE_PROJECT = gql`
  mutation deleteProject($id: ID) {
    deleteProject(id: $id) {
      project {
        id
      }
    }
  }
`;

const MOVE_PROJECT = gql`
  mutation moveProject($fromProjectId: ID, $toProjectId: ID) {
    moveProject(fromProjectId: $fromProjectId, toProjectId: $toProjectId) {
      project {
        id
        name
      }
    }
  }
`;

const UPDATE_PROJECT = gql`
  mutation updateProject($data: ProjectInputType, $id: ID) {
    updateProject(data: $data, id: $id) {
      project {
        id
        name
      }
    }
  }
`;

const ADD_ECOSYSTEM = gql`
  mutation addEcosystem($name: String!, $title: String, $description: String) {
    addEcosystem(name: $name, title: $title, description: $description) {
      ecosystem {
        id
      }
    }
  }
`;

const UPDATE_ECOSYSTEM = gql`
  mutation updateEcosystem($data: EcosystemInputType, $id: ID) {
    updateEcosystem(data: $data, id: $id) {
      ecosystem {
        id
      }
    }
  }
`;

const addProject = (apollo, data) => {
  const response = apollo.mutate({
    mutation: ADD_PROJECT,
    variables: {
      ecosystemId: data.ecosystemId,
      name: data.name,
      title: data.title,
      parentId: data.parentId
    }
  });
  return response;
};

const deleteProject = (apollo, id) => {
  const response = apollo.mutate({
    mutation: DELETE_PROJECT,
    variables: {
      id: id
    }
  });
  return response;
};

const moveProject = (apollo, fromProjectId, toProjectId) => {
  const response = apollo.mutate({
    mutation: MOVE_PROJECT,
    variables: {
      fromProjectId: fromProjectId,
      toProjectId: toProjectId
    }
  });
  return response;
};

const updateProject = (apollo, data, id) => {
  const response = apollo.mutate({
    mutation: UPDATE_PROJECT,
    variables: {
      data: data,
      id: id
    }
  });
  return response;
};

const addEcosystem = (apollo, data) => {
  const response = apollo.mutate({
    mutation: ADD_ECOSYSTEM,
    variables: {
      name: data.name,
      title: data.title,
      description: data.description
    }
  });
  return response;
};

const updateEcosystem = (apollo, data, id) => {
  const response = apollo.mutate({
    mutation: UPDATE_ECOSYSTEM,
    variables: {
      data: data,
      id: id
    }
  });
  return response;
};

export {
  addProject,
  deleteProject,
  moveProject,
  updateProject,
  addEcosystem,
  updateEcosystem
};
