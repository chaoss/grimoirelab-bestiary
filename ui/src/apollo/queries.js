import gql from "graphql-tag";

const projectFragment = gql`
  fragment projectFields on ProjectType {
    id
    name
    title
    ecosystem {
      id
      name
    }
    parentProject {
      name
      parentProject {
        name
        parentProject {
          name
        }
      }
    }
  }
`;

const GET_ECOSYSTEMS = gql`
  query GetEcosystems($pageSize: Int, $page: Int) {
    ecosystems(pageSize: $pageSize, page: $page) {
      entities {
        id
        name
        title
        description
        projectSet {
          ...projectFields
          subprojects {
            ...projectFields
            subprojects {
              ...projectFields
              subprojects {
                ...projectFields
              }
            }
          }
        }
      }
      pageInfo {
        totalResults
      }
    }
  }
  ${projectFragment}
`;

const GET_BASIC_PROJECT_INFO = gql`
  query GetBasicInfo($pageSize: Int, $page: Int, $filters: ProjectFilterType) {
    projects(pageSize: $pageSize, page: $page, filters: $filters) {
      entities {
        id
        title
      }
    }
  }
`;

const GET_PROJECTS = gql`
  query GetProjects($pageSize: Int, $page: Int, $filters: ProjectFilterType) {
    projects(pageSize: $pageSize, page: $page, filters: $filters) {
      entities {
        ...projectFields
        subprojects {
          ...projectFields
          subprojects {
            ...projectFields
            subprojects {
              ...projectFields
            }
          }
        }
      }
      pageInfo {
        page
        numPages
      }
    }
  }
  ${projectFragment}
`;

const GET_PROJECT_BY_NAME = gql`
  query getProjectById($filters: ProjectFilterType) {
    projects(pageSize: 1, page: 1, filters: $filters) {
      entities {
        ...projectFields
        subprojects {
          ...projectFields
          subprojects {
            ...projectFields
            subprojects {
              ...projectFields
            }
          }
        }
      }
    }
  }
  ${projectFragment}
`;

const getEcosystems = (apollo, pageSize, page) => {
  const response = apollo.query({
    query: GET_ECOSYSTEMS,
    variables: {
      pageSize,
      page
    },
    fetchPolicy: "no-cache"
  });
  return response;
};

const GetBasicProjectInfo = (apollo, pageSize, page, filters) => {
  const response = apollo.query({
    query: GET_BASIC_PROJECT_INFO,
    variables: {
      pageSize,
      page,
      filters
    }
  });
  return response;
};

const getProjects = (apollo, pageSize, page, filters) => {
  const response = apollo.query({
    query: GET_PROJECTS,
    variables: {
      pageSize,
      page,
      filters
    }
  });
  return response;
};

const getProjectByName = (apollo, name, ecosystemId) => {
  const response = apollo.query({
    query: GET_PROJECT_BY_NAME,
    variables: {
      filters: {
        ecosystemId: ecosystemId,
        name: name
      }
    },
    fetchPolicy: "no-cache"
  });
  return response;
};

export { getEcosystems, GetBasicProjectInfo, getProjects, getProjectByName };
