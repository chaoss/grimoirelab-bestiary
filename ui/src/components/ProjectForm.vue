<template>
  <v-form ref="form">
    <v-row>
      <v-col cols="4">
        <v-text-field
          v-model="title"
          label="Project title"
          :rules="validations.required"
          outlined
          dense
          @change="suggestName"
        ></v-text-field>
      </v-col>
    </v-row>
    <v-row>
      <v-col cols="4">
        <v-text-field
          v-model="name"
          label="Project name"
          :rules="validations.required"
          outlined
          dense
          @input="touchedName = true"
        ></v-text-field>
      </v-col>
    </v-row>
    <v-row>
      <v-col cols="4">
        <v-autocomplete
          v-model="parentId"
          :items="projects"
          label="Parent project (optional)"
          item-text="title"
          item-value="id"
          clearable
          outlined
          dense
          @click.once="loadParentProjects"
        >
          <template v-slot:no-data>
            <v-list-item>
              <v-list-item-content>
                <v-list-item-title class="text--disabled">
                  No available projects
                </v-list-item-title>
              </v-list-item-content>
            </v-list-item>
          </template>
        </v-autocomplete>
      </v-col>
    </v-row>
    <v-row>
      <v-col cols="4" class="d-flex justify-end">
        <v-btn color="primary" depressed @click="save">
          Save
        </v-btn>
      </v-col>
    </v-row>
  </v-form>
</template>

<script>
export default {
  name: "ProjectForm",
  props: {
    ecosystemId: {
      type: Number,
      required: false
    },
    getProjects: {
      type: Function,
      required: true
    },
    addProject: {
      type: Function,
      required: true
    },
    parent: {
      type: Object,
      required: false,
      default: null
    }
  },
  data() {
    return {
      name: null,
      title: null,
      parentId: null,
      touchedName: false,
      projects: [],
      validations: {
        required: [value => !!value || "Required"]
      }
    };
  },
  methods: {
    suggestName(value) {
      if (value && !this.touchedName) {
        this.name = value
          .trim()
          .replace(/\s+/g, "-")
          .toLowerCase();
      }
    },
    async loadParentProjects() {
      const response = await this.getProjects(Number(this.ecosystemId));
      if (response) {
        this.projects = response;
      }
    },
    async save() {
      if (!this.$refs.form.validate()) {
        return;
      }
      const data = {
        name: this.name.trim(),
        title: this.title,
        parentId: this.parentId ? Number(this.parentId) : null,
        ecosystemId: this.ecosystemId
      };
      const response = await this.addProject(data);
      if (response) {
        const projectName = response.name;
        this.$router.push({
          path: `/ecosystem/${this.ecosystemId}/project/${projectName}`
        });
      }
    }
  },
  created() {
    if (this.parent) {
      this.loadParentProjects();
      this.parentId = this.parent.id;
    }
  }
};
</script>
