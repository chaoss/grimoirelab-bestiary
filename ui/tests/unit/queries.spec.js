import { shallowMount } from "@vue/test-utils";
import Vue from "vue";
import Vuetify from "vuetify";
import * as Queries from "@/apollo/queries";
import ProjectForm from "@/components/ProjectForm";

Vue.use(Vuetify);

describe("ProjectsForm queries", () => {
  const getProjectsResponse = {
    data: {
      projects: {
        entities: [{ id: "2", title: "Test", __typename: "ProjectType" }]
      }
    }
  };

  test("Mock query for getProjects", async () => {
    const querySpy = spyOn(Queries, "getProjects");
    const query = jest.fn(() => Promise.resolve(getProjectsResponse));
    const wrapper = shallowMount(ProjectForm, {
      Vue,
      mocks: {
        $apollo: {
          query
        }
      },
      propsData: {
        ecosystemId: 1,
        getProjects: Queries.getProjects,
        addProject: () => {}
      }
    });
    await wrapper.vm.loadParentProjects();

    expect(querySpy).toHaveBeenCalledWith(1);
    expect(wrapper.element).toMatchSnapshot();
  });
});
