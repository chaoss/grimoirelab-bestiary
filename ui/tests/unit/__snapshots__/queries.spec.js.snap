// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ecosystem queries Mock query for getEcosystemByID 1`] = `<!---->`;

exports[`ProjectsForm queries Mock query for getProjects 1`] = `
<v-form-stub>
  <v-row-stub
    tag="div"
  >
    <v-col-stub
      cols="4"
      tag="div"
    >
      <v-text-field-stub
        backgroundcolor=""
        clearicon="$clear"
        dense="true"
        errorcount="1"
        errormessages=""
        label="Project title"
        loaderheight="2"
        messages=""
        outlined="true"
        rules="value => !!value || \\"Required\\""
        successmessages=""
        type="text"
      />
    </v-col-stub>
  </v-row-stub>
   
  <v-row-stub
    tag="div"
  >
    <v-col-stub
      cols="4"
      tag="div"
    >
      <v-text-field-stub
        backgroundcolor=""
        clearicon="$clear"
        dense="true"
        errorcount="1"
        errormessages=""
        label="Project name"
        loaderheight="2"
        messages=""
        outlined="true"
        rules="value => !!value || \\"Required\\""
        successmessages=""
        type="text"
      />
    </v-col-stub>
  </v-row-stub>
   
  <v-row-stub
    tag="div"
  >
    <v-col-stub
      cols="4"
      tag="div"
    >
      <v-autocomplete-stub
        allowoverflow="true"
        appendicon="$dropdown"
        backgroundcolor=""
        clearable="true"
        clearicon="$clear"
        dense="true"
        errorcount="1"
        errormessages=""
        filter="function _default(item, queryText, itemText) {
        return itemText.toLocaleLowerCase().indexOf(queryText.toLocaleLowerCase()) > -1;
      }"
        itemcolor="primary"
        itemdisabled="disabled"
        items=""
        itemtext="title"
        itemvalue="id"
        label="Parent project (optional)"
        loaderheight="2"
        menuprops="[object Object]"
        messages=""
        nodatatext="$vuetify.noDataText"
        outlined="true"
        rules=""
        successmessages=""
        type="text"
        valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date && b instanceof Date && a.getTime() !== b.getTime()) {
    // If the values are Date, compare them as timestamps
    return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}"
      />
    </v-col-stub>
  </v-row-stub>
   
  <v-row-stub
    tag="div"
  >
    <v-col-stub
      class="d-flex justify-end"
      cols="4"
      tag="div"
    >
      <v-btn-stub
        activeclass=""
        color="primary"
        depressed="true"
        tag="button"
        type="button"
      >
        
        Save
      
      </v-btn-stub>
    </v-col-stub>
  </v-row-stub>
</v-form-stub>
`;
